### PR TITLE
US019: Add view list of free slots

### PR DESCRIPTION
**Pull request by SamDotNetDev**

### CODE REVIEW

File: ParkingZoneApp/Models/Reservation.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+﻿using ParkingZoneApp.Data.Migrations;
+using System.ComponentModel.DataAnnotations;
+
+namespace ParkingZoneApp.Models
+{
+    public class Reservation
+    {
+        [Key]
+        [Required]
+        public int ReservationId { get; set; }
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Models/Reservation.cs]: image

File: ParkingZoneApp/Models/Reservation.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        public int Duration { get; set; }
+
+        [Required]
+        public int ParkingSlotId { get; set; }
+
+        [Required]
+        public virtual ParkingSlot ParkingSlot { get; set; }
+
+        [Required]
+        public virtual ApplicationUser User { get; set; }
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Models/Reservation.cs:28]: also add user Id

File: ParkingZoneApp/Program.cs [hunk 3]
```diff
@@ -56,11 +58,15 @@ public static void Main(string[] args)
             app.MapControllerRoute(
                 name: "Admin",
                 pattern: "{area:exists}/{controller=ParkingZone}/{action=Index}/{id?}");
-
+
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Program.cs]: space

File: ParkingZoneApp/Services/IParkingSlotService.cs [hunk 4]
```diff
@@ -1,11 +1,12 @@
...
-using ParkingZoneApp.Repositories;
 
 namespace ParkingZoneApp.Services
 {
     public interface IParkingSlotService : IService<ParkingSlot>
     {
         public IEnumerable<ParkingSlot> GetByParkingZoneId(int parkingZoneId);
-        public bool ParkingSlotExits(int ParkingZoneId, int ParkingSlotNumber);
+        public bool ParkingSlotExists(int parkingZoneId, int parkingSlotNumber);
+        public bool SlotFree(ParkingSlot slot, DateTime starTime, int duration);
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Services/IParkingSlotService.cs]: when we read method name, we should understand what it does. in your case its little bit weird. (naming convention)

File: ParkingZoneApp/Services/ParkingSlotService.cs [hunk 5]
```diff
@@ -15,12 +16,29 @@ public IEnumerable<ParkingSlot> GetByParkingZoneId(int parkingZoneId)
...
         }
+
+        public IEnumerable<ParkingSlot> GetFreeByParkingZoneIdAndPeriod(int parkingZoneId, DateTime startTime, int duration)
+        {
+            var slots = _repository.GetAll()
+                .Where(x => x.ParkingZoneId == parkingZoneId && x.IsAvailableForBooking);
+            return slots;
+        }
+
+        public bool SlotFree(ParkingSlot slot, DateTime startTime, int duration)
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Services/ParkingSlotService.cs]: so you wrote logic. I have a question for you where did you use it ?!

File: ParkingZoneApp/Services/ParkingSlotService.cs [hunk 6]
```diff
@@ -15,12 +16,29 @@ public IEnumerable<ParkingSlot> GetByParkingZoneId(int parkingZoneId)
...
-                x.Number == ParkingSlotNumber);
+                .Where(x => x.ParkingZoneId == parkingZoneId &&
+                x.Number == parkingSlotNumber);
             return ParkingSlots.Count() == 0 ? false : true;
         }
+
+        public IEnumerable<ParkingSlot> GetFreeByParkingZoneIdAndPeriod(int parkingZoneId, DateTime startTime, int duration)
+        {
+            var slots = _repository.GetAll()
+                .Where(x => x.ParkingZoneId == parkingZoneId && x.IsAvailableForBooking);
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Services/ParkingSlotService.cs]: this logic does not return free slots but only all slots which depend on specific parking zone

File: ParkingZoneApp/Views/Reservation/FreeSlots.cshtml [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+@if(Model.ParkingSlots is not null && Model.ParkingSlots.Count() > 0)
+{
+    <h1>Free Slots in @ViewData["ParkingZoneName"]</h1>
+    <hr />
+    <button class="button-34" onclick="location.href='@Url.Action("FreeSlots","Reservation")'">Back</button>
+    <table class="table">
+        <thead>
+            <tr>
+                <th>Number</th>
+                <th>Category</th>
```
Comment (abdullokhtolibjonov) [inline ParkingZoneApp/Views/Reservation/FreeSlots.cshtml]: lets use MVC way html

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

File: ParkingZoneApp/Services/ParkingSlotService.cs (hunk 6 above)
Comment (SamanKingDotNetDeveloper) [reply ParkingZoneApp/Services/ParkingSlotService.cs:30]: image
I can not see it in my visual studio is is showing normal

File: ParkingZoneApp/Program.cs (hunk 3 above)
Comment (SamDotNetDev) [reply ParkingZoneApp/Program.cs]: image
In visual studio it is not showing any spaces

File: ParkingZoneApp/Controllers/ReservationController.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        private readonly IReservationService _reservationService;
+        private readonly IParkingSlotService _slotService;
+        private readonly IParkingZoneService _zoneService;
+
+        public ReservationController(IReservationService reservationService ,IParkingSlotService slotService, IParkingZoneService zoneService)
+        {
+            _reservationService = reservationService;
+            _slotService = slotService;
+            _zoneService = zoneService;
+
```
Comment (dilmurod-rozikov) [inline ParkingZoneApp/Controllers/ReservationController.cs]: Empty line :)

File: ReservationTests/ModelTests/FreeSlotsVMValidationTests.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        public static IEnumerable<object[]> TestData =>
+        new List<object[]>
+        {
+            new object[] { DateTime.Now, 1, 1, null, null, true },
+        };
+
+        [Theory]
+        [MemberData(nameof(TestData))]
+        public void GivenItemToBeValidated_WhenCreatingCreateVM_ThenValidationIsPerformed(DateTime StartTime, int Duration, int ParkingZoneId, SelectList ParkingZones, IEnumerable<ListItemVM> ParkingSlots, bool expectedValidationResult)
```
Comment (dilmurod-rozikov) [inline ReservationTests/ModelTests/FreeSlotsVMValidationTests.cs]: naming is not correct maybe _WhenCreatingFreeSlotsVM

**Review by dilmurod-rozikov — APPROVED** [review]

**Review by dilmurod-rozikov — APPROVED** [review]

File: ParkingZoneApp/Program.cs (hunk 3 above)
Comment (abdullokhtolibjonov) [reply ParkingZoneApp/Program.cs]: just delete this row and add again with enter

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

File: Tests/ReservationTests/Controller/ReservationControllerTests.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+            //Act
+            var result = _controller.FreeSlots(vm);
+
+            //Assert
+            Assert.NotNull(result);
+            var model = Assert.IsType<ViewResult>(result).Model;
+            Assert.IsType<FreeSlotsVM>(model);
+            _zoneService.Verify(x => x.GetAll(), Times.Once);
+            _slotService.Verify(x => x.GetFreeByParkingZoneIdAndPeriod(vm.ParkingZoneId, vm.StartTime, vm.Duration), Times.Once);
```
Comment (abdullokhtolibjonov) [inline Tests/ReservationTests/Controller/ReservationControllerTests.cs:72]: JSON SERIALIZER

Comment (abdullokhtolibjonov) [reply Tests/ReservationTests/Controller/ReservationControllerTests.cs:72]: compare expected and result

**Review by abdullokhtolibjonov — CHANGES_REQUESTED** [review]

**Review by dilmurod-rozikov — APPROVED** [review]

**Review by abdullokhtolibjonov — APPROVED** [review]

**Pull request merged**